### PR TITLE
Parallelize tests, remove data races.

### PR DESCRIPTION
**Pull request by tmc**
This changes the tests to run in parallel and goes through and refactors all the code to remove detected races

Further work here would be to for parallelism within a test case (likely using t.Run()) which will make #52 more relevant.

This is groundwork to support multiple queries within the same address space, namely for #42

### CODE REVIEW

File: src/sybil/query_spec.go [hunk 1]
```diff
@@ -80,7 +80,7 @@ type Aggregation struct {
 	opID     int
 	Name     string
 	nameID   int16
-	HistType string
+	HistType HistogramType
```
Comment (okayzed) [inline src/sybil/query_spec.go]: you are breaking my query cache here. i need a sybil command to clean caches (or expire cached results after X time period)

File: src/sybil/column_store_test.go [hunk 2]
```diff
@@ -98,7 +106,7 @@ func TestColumnStoreFileNames(t *testing.T) {
 		t.Error("Found other records than rowblock")
 	}
 
-	nt.DigestRecords(minFilesToDigest)
+	nt.DigestRecords(minFilesToDigest, digestSpec)
```
Comment (okayzed) [inline src/sybil/column_store_test.go]: should minFilesToDigest be pushed into digestSpec?

File: src/cmd/cmd_trim.go [hunk 3]
```diff
@@ -61,10 +61,11 @@ func RunTrimCmdLine() {
 	loadSpec.TimeColumn = *flags.TIME_COL
 
 	trimSpec := sybil.TrimSpec{}
+	trimSpec.TimeColumn = *flags.TIME_COL
 	trimSpec.DeleteBefore = int64(*DELETE_BEFORE)
 	trimSpec.MBLimit = int64(*MB_LIMIT)
 
-	toTrim := t.TrimTable(flags, &trimSpec)
+	toTrim := t.TrimTable(&trimSpec)
```
Comment (okayzed) [inline src/cmd/cmd_trim.go]: note to self: do we have trimming tests?

File: src/cmd/cmd_query.go [hunk 4]
```diff
@@ -84,8 +84,13 @@ func RunQueryCmdLine() {
 		return
 	}
 
+	printConfig := sybil.PrintConfig{
```
Comment (okayzed) [inline src/cmd/cmd_query.go]: not printSpec? ;)

File: src/sybil/column_store_test.go [hunk 5]
```diff
@@ -184,7 +184,7 @@ func TestBigIntColumns(t *testing.T) {
 
 	unloadTestTable(tableName)
 
-	nt.Options.SAMPLES = true
+	OPTS.SAMPLES = true
```
Comment (okayzed) [inline src/sybil/column_store_test.go]: what's going on with OPTS? still exists or not?

File: src/cmd/cmd_query.go [hunk 6]
```diff
@@ -286,13 +291,11 @@ func RunQueryCmdLine() {
...
-			t.LoadAndQueryRecords(flags, &loadSpec, &querySpec)
+		t.LoadAndQueryRecords(flags, &loadSpec, &querySpec)
 
-			end := time.Now()
-			sybil.Debug("LOAD AND QUERY RECORDS TOOK", end.Sub(start))
-			querySpec.PrintResults(flags)
-		}
+		end := time.Now()
+		sybil.Debug("LOAD AND QUERY RECORDS TOOK", end.Sub(start))
+		querySpec.PrintResults(*flags.OP, printConfig)
```
Comment (okayzed) [inline src/cmd/cmd_query.go]: why flags.OP not in printConfig?

File: src/sybil/table_query.go [hunk 7]
```diff
@@ -216,10 +216,10 @@ func (t *Table) LoadAndQueryRecords(flags *FlagDefs, loadSpec *LoadSpec, querySp
 				}
 			}(t)
 
-			if t.Options.SAMPLES {
+			if OPTS.SAMPLES {
 				wg.Wait()
 
-				if count > querySpec.Limit {
+				if querySpec != nil && count > querySpec.Limit {
```
Comment (okayzed) [inline src/sybil/table_query.go]: why was this added? (i'm curious about the context that led to needing this check)

File: src/sybil/record_slab.go [hunk 8]
```diff
@@ -2,9 +2,9 @@ package sybil
 
 import "time"
 
-func (tb *TableBlock) allocateRecords(loadSpec *LoadSpec, info SavedColumnInfo, loadRecords bool) RecordList {
+func (tb *TableBlock) allocateRecords(flags *FlagDefs, loadSpec *LoadSpec, info SavedColumnInfo, loadRecords bool) RecordList {
```
Comment (okayzed) [inline src/sybil/record_slab.go]: the table doesn't have flags set on it (because a table will eventually be shared between queries) so we need to pass this flags object around everywhere? seems ungainly. one possible solution is a cursor: cursor wraps a table + flags and exposes API, so we don't thread flags around.

File: src/sybil/row_store.go [hunk 9]
```diff
@@ -129,20 +129,20 @@ func (t *Table) LoadRecordsFromLog(filename string) RecordList {
...
 
 }
 
-func (t *Table) AppendRecordsToLog(records RecordList, blockname string) {
+func (t *Table) AppendRecordsToLog(flags *FlagDefs, records RecordList, blockname string) {
 	if len(records) == 0 {
 		return
 	}
 
 	// TODO: fix this up, so that we don't
```
Comment (okayzed) [inline src/sybil/row_store.go]: so that we don't what?! damnit, past me - what warning were you trying to give us?

File: src/sybil/query_spec.go (hunk 1 above)
Comment (tmc) [reply src/sybil/query_spec.go]: I can back out that type change.

File: src/sybil/column_store_test.go (hunk 2 above)
Comment (tmc) [reply src/sybil/column_store_test.go]: I did this at first but it felt a bit awkward as it's returned from IngestRecords which accepts a digestSpec.

File: src/cmd/cmd_trim.go (hunk 3 above)
Comment (tmc) [reply src/cmd/cmd_trim.go]: initial coverage check seems to indicate TrimTable is not covered in test.

File: src/cmd/cmd_query.go (hunk 4 above)
Comment (tmc) [reply src/cmd/cmd_query.go]: :)

File: src/sybil/column_store_test.go (hunk 5 above)
Comment (tmc) [reply src/sybil/column_store_test.go]: I had intended on removing it completely, I'll revise and do so.

File: src/cmd/cmd_query.go (hunk 6 above)
Comment (tmc) [reply src/cmd/cmd_query.go]: It wasn't used in all the places so I wasn't sure it belonged.

Comment (okayzed) [reply src/cmd/cmd_query.go]: "better stuffed than sorry"

File: src/sybil/table_query.go (hunk 7 above)
Comment (tmc) [reply src/sybil/table_query.go]: I think it was added before the outer nil check came into play, not necessary now.

File: src/sybil/record_slab.go (hunk 8 above)
Comment (tmc) [reply src/sybil/record_slab.go]: IMO virtually any flag passing should be turned into better parameters

Comment (tmc) [reply src/sybil/record_slab.go]: in this case a later commit removes this param (33c69fe4e869319a6c2d356ee2971c02393e6b81 at time of writing)

File: src/sybil/query_spec.go (hunk 1 above)
Comment (tmc) [reply src/sybil/query_spec.go]: As discussed I think I'll be suggesting a alternate method for cache key generation that would have the benefit of being stable in the case of this type of change.

**Comment by tmc** [comment]
So this got a little big and unwieldy. I opened https://github.com/logv/sybil/pull/61 to be a small part of this work and think it will be easier to review in smaller batches.
